src/sql/workbench/contrib/notebook/test/browser/notebookEditor.test.ts

```diff
@@ -422,11 +376,358 @@ suite('Test class NotebookEditor', () => {
+			}
+		}
+	});
+
+	for (const currentMatch of [null, new NotebookRange(<ICellModel>{}, 1, 1, 1, 1)]) {
+		const searchString = getRandomString(1, 10); // a random string 1 to 10 characters long.
+		for (const modelFindExpression of [searchString, getRandomString(1, 10)]) {
+			for (const matchCase of [false, true]) {
+				for (const wholeWord of [false, true]) {
+					for (const findMatches of [[], [new NotebookFindMatch(currentMatch || <NotebookRange>{}, null)]]) {
+						test(`Verifies _onFindStateChange callback when searchString='${searchString}', currentMatch='${currentMatch}', findExpression='${modelFindExpression}', matchCase='${matchCase}', wholeWord='${wholeWord}', findMatches='${findMatches}'`, async () => {
+							const { findReplaceStateChangedEvent, notebookFindModelMock, findDecorationsMock, notebookEditor } = await findStateChangeSetup(instantiationService, workbenchThemeService, notebookService, untitledNotebookInput, modelFindExpression, currentMatch, searchString, wholeWord, matchCase, null, findMatches);
+							await notebookEditor['_onFindStateChange'](findReplaceStateChangedEvent);
+							if (currentMatch) {
+								assert.strictEqual(notebookEditor.getPosition(), currentMatch, `position must be set to the same NotebookRange that we set for '_currentMatch' property of notebookEditor`);
+							}
+							if (searchString === modelFindExpression && currentMatch && !matchCase && !wholeWord) {
+								notebookFindModelMock.verify(x => x.find(
+									TypeMoq.It.isAnyString(),
+									TypeMoq.It.isAny(),
+									TypeMoq.It.isAny(),
+									TypeMoq.It.isAnyNumber()
+								), TypeMoq.Times.never());
+								findDecorationsMock.verify(x => x.set(TypeMoq.It.isAny(), TypeMoq.It.isAny()), TypeMoq.Times.never());
+							} else {
+								notebookFindModelMock.verify(x => x.find(
+									TypeMoq.It.isAnyString(),
+									TypeMoq.It.isAny(),
+									TypeMoq.It.isAny(),
+									TypeMoq.It.isAnyNumber()
+								), TypeMoq.Times.once());
+								assert.strictEqual(notebookEditor.notebookFindModel.findExpression, searchString, `findExpression should get set to the provided searchString:${searchString}`);
+								if (!currentMatch && notebookEditor.notebookFindModel.findMatches.length === 0) {
+									notebookFindModelMock.verify(x => x.clearFind(), TypeMoq.Times.once());
+									findDecorationsMock.verify(x => x.set(TypeMoq.It.isAny(), TypeMoq.It.isAny()), TypeMoq.Times.never());
+									assert.strictEqual(notebookEditor.notebookFindModel.findArray?.length ?? 0, 0, 'The find array should be cleared or undefined if there were no findMatches');
+								} else {
+									findDecorationsMock.verify(x => x.set(TypeMoq.It.isAny(), TypeMoq.It.isAny()), TypeMoq.Times.once());
+								}
+							}
+							findDecorationsMock.verify(x => x.clearDecorations(), TypeMoq.Times.once());
+						});
+					}
+				}
 			}
 		}
+	}
+
+	for (const searchString of ['', undefined]) {
```

Comment:
It would be a lot cleaner if you just made test functions that took these in as parameters and called them that way. These nested loops are getting a little much

Reply:
I will make a test function. From what I can see that won't remove the nested loops but will make the code more readable to sit is a good thing to do.

Reply:
done